Add FAQ to documentation (#1212)
spockframework/spock: docs/faq.adoc

= Frequently Asked Questions (FAQ)
include::include.adoc[]

== What are Spock's optional dependencies used for?

Spock runs fine without optional dependencies, hence the term _optional_. So what are they used for?

  * Objenesis (artifact `org.objenesis:objenesis`) is required for mocking classes without default constructors.

>   * Either CGLIB (artifact `cglib:cglib-nodep`) or ByteBuddy (artifact `net.bytebuddy:byte-buddy`) are required for

Reviewer: ```suggestion
  * Either ByteBuddy (artifact `net.bytebuddy:byte-buddy`) or CGLIB (artifact `cglib:cglib-nodep`) are required for
```

    mocking classes and not just interfaces.

See also <<interaction_based_testing.adoc#_mocking_classes,Mocking Classes>>.

> == Why are there two options (CGLIB vs ByteBuddy) for mocking classes?

Reviewer: ```suggestion
== Why are there two options (ByteBuddy vs CGLIB) for mocking classes?
```

The main reason is backwards compatibility. CGLIB came first historically, ByteBuddy was added later. Both libraries
basically provide identical features from a Spock user's perspective.

If you are starting with Spock now, we recommend ByteBuddy as our preferred choice.

> As of today there are no plans to remove CGLIB anytime soon, because the maintenance effort is relatively low. So you
> may safely continue to use CGLIB if e.g. your tests use it in another context already. If you have no compelling reasons

Reviewer: ```suggestion
CGLIB is in legacy support mode, we will support it as long as the maintenance effort is relatively low. So you
may continue to use CGLIB for now if e.g. your tests use it in another context already. If you have no compelling reasons
```
